JuliaLowering/src/scope_analysis.jl

```diff
@@ -17,493 +17,405 @@ function NameKey(ex::SyntaxTree)
+                ctx.scope_layers[vk.layer].mod, Symbol(vk.name))
+                # special soft scope rules: existing global variables are assigned to
+                if ctx.enable_soft_scopes
+                    push!(ctx.defined_toplevel_globals, vk)
+                    declare_in_scope!(ctx, top_scope(ctx), ex, :global)
+                else
+                    declare_in_scope!(ctx, scope, ex, :local; is_ambiguous_local=true)
                 end
+            else
+                declare_in_scope!(ctx, scope, ex, :local)
             end
-        end
-    end
-
-    if !in_toplevel_thunk
-        for (varkey,_) in assignments
-            id = haskey(var_ids, varkey) ? var_ids[varkey] : lookup_var(ctx, varkey)
-            binfo = lookup_binding(ctx, id)
-            if binfo.kind !== :global
-                if !has_lambda_binding(lambda_bindings, id)
-                    # Assigned vars from a scope *outside* the current lambda are captured
-                    init_lambda_binding(lambda_bindings, id, is_captured=true)
-                    update_binding!(ctx, id; is_captured=true)
-                end
+        elseif b.kind === :global
+            # This corner case ignores global decls in the top-level scope
+            if is_toplevel_thunk ||
+                !isnothing(resolve_name(ctx, ex; exclude_toplevel_globals=true)) ||
+                (ctx.enable_soft_scopes && scope.is_permeable &&
+                vk in ctx.defined_toplevel_globals)
+                # assign-existing-global in all other scenarios: (1) at top
+                # level, (2) if this is an explicit global that isn't at top
+                # level, or (3) if the soft scope exception applies
+                push!(ctx.defined_toplevel_globals, vk)
+            else
+                declare_in_scope!(ctx, scope, ex, :local)
             end
+        elseif b.kind === :static_parameter
+            throw(LoweringError(ex, "cannot overwrite a static parameter"))
+        elseif b.kind === :local || b.kind === :argument
+            # unambiguous assignment to existing variable
+            record_lambda_var!(ctx, scope, b, true)
         end
     end
 
-    return ScopeInfo(is_toplevel_global_scope, in_toplevel_thunk, is_soft_scope,
-                     is_hard_scope, var_ids, lambda_bindings)
+    return scope
 end
 
 function add_local_decls!(ctx, stmts, srcref, scope)
     # Add local decls to start of block so that closure conversion can
     # initialize if necessary.
-    for id in sort!(collect(values(scope.var_ids)))
-        binfo = lookup_binding(ctx, id)
+    for id in sort!(collect(values(scope.vars)))
+        binfo = get_binding(ctx, id)
         if binfo.kind == :local
             push!(stmts, @ast ctx srcref [K"local" binding_ex(ctx, id)])
         end
     end
 end
 
-function _resolve_scopes(ctx, ex::SyntaxTree)
+function _resolve_scopes(ctx, ex::SyntaxTree, @nospecialize(scope))
```

Comment:
```suggestion
function _resolve_scopes(ctx, ex::SyntaxTree, @nospecialize(scope::Union{Nothing,ScopeInfo}))
```

IIUC you assume this, so best to mark it